### PR TITLE
Balance for Rebels 1.18

### PR DESCRIPTION
**Pull request by Hejnewar**
Level 1: 
Fighter - xp cost changed from 40 to 37.
Merman Hunter - xp changed from 35 to 33.
Scout - xp changed from 32 to 42.

Level 2: 
Hero - xp changed from 90 to 80.
Captain - xp changed from 90 to 85.
Marksman - cost changed from 31 to 34, xp changed from 80 to 80.
Ranger - xp changed from 90 to 85. 
Rider - ranged damage changed from 9 to 11, hp changed form 46 to 49, cost changed from 28 to 35, xp changed from 53 to 75.
Druid - cost changed from 27 to 25, xp changed from 80 to 85.
Sorceress - cost changed from 32 to 34, xp changed from 100 to 90. 
Elder Wose - cost changed from 27 to 28, xp changed from 100 to 80.
Merman Netcaster - cost changed from 27 to 26, xp changed from 85 to 54.
Merman Spearman - cost changed from 27 to 22, xp changed from 85 to 54.

Level 3:
Champion - melee damage changed from 9 to 8. added 10 melee accuracy, hp changed from 70 to 72, cost changed from 48 to 61.
Marshal - hp changed from 62 to 68, cost changed from 54 to 67.
Sharpshooter - ranged damage changed from 10 to 12. cost changed from 51 to 62.
Avenger - ranged damage changed from 10 to 11, hp changed from 55 to 59, cost changed from 53 to 66.
Outrider - ranged damage changed from 8 to 11, melee damage changed from 7 to 8,  hp changed form 57 to 60, mp from 10 to 11, cost changed from 43 to 58.
Shyde - hp changed from 46 to 51, melee damage changed from 6 to 7, ranged slow (entangle) changed from 6 to 8, ranged magical (thorns) changed from 8 to 14, cost changed from 52 to 69.
Enchantress - cost changed from 55 to 70, ranged slow damage changed from 5 to 7, ranged magical damage changed from 9 to 13, xp changed from 180 to 180.
Ancient Wose - cost changed from 48 to 50.
Entangler - cost changed from 46 to 42.
Javelineer - cost changed from 48 to 55.

Level 4:
Sylph - hp changed from 60 to 68, ranged slow damage changed from 6 to 7, ranged magical damage changed from 10 to 16, cost changed from 67 to 148 (this is wrong but as commited will fix latter).

### CODE REVIEW

**Review by knyghtmare — APPROVED** [review]

**Comment by Pentarctagon** [comment]
Looks like this broke the pathfinding WML unit tests, probably because of the Outrider's movement change.

**Comment by Hejnewar** [comment]
The real question is what can be done about that? Honestly tests breaking with unit mp changes doesnt sound great.

**Comment by stevecotton** [comment]
Having the pathfinding tests depend on the mainline units seems a reasonable tradeoff to me, given that a rebalance is a very rare event. There's documentation about how to update these tests:
https://github.com/wesnoth/wesnoth/blob/b50cf83f7e358ee11b094d129861f65a51bcdfa3/data/test/scenarios/wml_tests/ScenarioWML/EventWML/ActionWML/InternalActionsWML/characterize_pathfinding.cfg#L2-L14

**Comment by Pentarctagon** [comment]
It might be simplest to have a separate PR specifically for the movement changes. Also any changes to the elvish archer's or orcish grunt's hp or weapon damage would likely break a bunch of tests, since those are the two units used in nearly all test scenarios.

**Comment by Hejnewar** [comment]
Thats decent idea. If archer got changes it would be in this PR, and grunt doesnt get any changes, so at least thats something I guess.

**Comment by stevecotton** [comment]
Movement changes are probably the changes most likely to need campaign changes. For example, SoF S07:
https://github.com/wesnoth/wesnoth/blob/1efa0c8c54eb9247b93bcc9dc1ba96cc477b183f/data/campaigns/Sceptre_of_Fire/scenarios/7_Outriding_the_Outriders.cfg#L324-L329

I don't know how many other examples there are, I've just remembered that scenario.

**Comment by Hejnewar** [comment]
Well, honestly changes to rebels are probably the biggest in terms of numbers, because they were supposed to be more elite. I just did what I thought was in character for the unit line, but I can simply swap it for something else if needed.

File: data/core/units/elves/Champion.cfg [hunk 1]
```diff
@@ -24,8 +24,9 @@
         icon=attacks/greatsword-elven.png
         type=blade
         range=melee
-        damage=9
+        damage=8
         number=5
+        accuracy=10
```
Comment (cooljeanius) [inline data/core/units/elves/Champion.cfg:29]: Why modify accuracy directly in the attack like this, rather than adding a weapon special?

Comment (Hejnewar) [reply data/core/units/elves/Champion.cfg:29]: What is the purpose of this stat existing then? It exists so im using it. Not gonna do weapon specials for stuff that exists already as a stat.

Comment (cooljeanius) [reply data/core/units/elves/Champion.cfg:29]: well ok, does it show up anywhere in-game? How is the player supposed to know about it being there?

Comment (Hejnewar) [reply data/core/units/elves/Champion.cfg:29]: Why is this question for me? If players dont see unit stats then thats UI probelm. 

It is visible, not very well explained, but visible.

Comment (stevecotton) [reply data/core/units/elves/Champion.cfg:29]: There are no unit tests for `accuracy`, it's not used directly in mainline WML (there's some use of `increase_accuracy`, but not `accuracy`), and it's used exactly once in the entirety of Ageless Era. It has the C++ code to support it in the combat calculations and store_unit/unstore_unit, but that's about it.

If we're going to support it properly, it should have tests and UI, and there are already two issues about that (#2517 and #5860). However, as we're getting close to the release candidates for 1.18, it seems late in the development cycle to add those to the roadmap.

Using a weapon special instead would be a simple way to ensure it works well with the current UI, and at this point in the 1.18 cycle that seems the best choice to me.

Comment (knyghtmare) [reply data/core/units/elves/Champion.cfg:29]: >  If players dont see unit stats then thats UI problem.

I agree with Hejnewar here. The UI should be enhanced to augment the UX of the game and I dont see this is a argumentative blocker for the balance patch. Also, the accuracy of +10% is shown in the sidebar under the unit's attack as a `+10%/0%` which means +10% to CTH on attack, and 0% to enemy CTH on defense.

Comment (Hejnewar) [reply data/core/units/elves/Champion.cfg:29]: So, because no one uses it, it was never a feature that was made stable and its usage is discouraged because it is unstable? Since I wanted to use it in many places, there was an issue for few years. It wasnt made any less bad. Now I want to use it. 

And well I will use it anyway so maybe that issue is finally resolved.

Comment (stevecotton) [reply data/core/units/elves/Champion.cfg:29]: If we were at 1.17.5 or 1.19.5, then I'd agree that using it is the best way to get the issues tested and resolved.

But we're about 3 months before string-freeze with other UI issues that need fixing, and still working on getting the event system and ability filter changes tested. I hope that the accuracy feature is stable, but if it isn't, then I doubt it will be fixed for the 1.18 branch.

Comment (Pentarctagon) [reply data/core/units/elves/Champion.cfg:29]: What's the problem with the current UI for it?

Also, "you can't use features that have no unit tests" isn't a requirement that I'm aware of? It would certainly be good for a unit test to be added for it, but as long as Hejnewar has tested it himself and can confirm that it does work the way it's supposed to, then IMO that should be good enough.

Comment (Hejnewar) [reply data/core/units/elves/Champion.cfg:29]: It just doesnt say what it does, if we threat it like any other stat that doesnt need explaining thats fine but right now acuraccy x / parry y is all it says. We can assume its fine but usually there is some confusion around it.

Comment (stevecotton) [reply data/core/units/elves/Champion.cfg:29]: Changing the tooltip in #5860 to say `<b>Accuracy:</b> +10% bonus to chance to hit` would solve a lot of my concerns about the UI. Not difficult, but without @cooljeanius flagging this issue it wouldn't have happened before the string-freeze. I'm still not sure that it will happen in time for the string freeze.

> Also, "you can't use features that have no unit tests" isn't a requirement that I'm aware of?

It's more a general feel of how many of the edge cases have been tested, how it interacts with other abilities that change the chance-to-hit stats, for SP how it interacts with the PRNG options. The existence of unit tests is reassuring, the amount that it gets used in Mainline + Ageless is my main measure for guessing how many of the edge cases are known.

For PRs, my sad lesson from experience is to assume that testing was minimal unless someone say otherwise.

Comment (Pentarctagon) [reply data/core/units/elves/Champion.cfg:29]: @Hejnewar have you tested using the Champion with the `accuracy` attribute added? Or @ProditorMagnus have there been any issues with it in Ageless Era?

Comment (ProditorMagnus) [reply data/core/units/elves/Champion.cfg:29]: Rashy Era uses it out of actively developed source eras. With 4.30 it was also added to several other units for balancing. I would encourage tooltip change too, but values are visible from attack dialog so that should not be a blocker. @andrzejskutela uses it for boss units too.

Comment (stevecotton) [reply data/core/units/elves/Champion.cfg:29]: I was looking at Ageless 4.28. I confirm, I'm a lot happier about this PR after looking at the current Ageless.

Comment (ProditorMagnus) [reply data/core/units/elves/Champion.cfg:29]: Getting it to core unit should also help to speed up getting it to wmlunits.

Comment (cooljeanius) [reply data/core/units/elves/Champion.cfg:29]: > Getting it to core unit should also help to speed up getting it to wmlunits.

Is there a separate issue for that?

Comment (andrzejskutela) [reply data/core/units/elves/Champion.cfg:29]: for the record, since my name was mentioned
[accuracy + parry visual](https://ibb.co/QXYJR6y)
[hover](https://ibb.co/x5Tw6NV)

Comment (ProditorMagnus) [reply data/core/units/elves/Champion.cfg:29]: Shouldnt be.

Comment (Hejnewar) [reply data/core/units/elves/Champion.cfg:29]: Well for now I will merge it, if there is still some discussion needed it can be moved to an issue and if something is really wrong I can always change the change.

Comment (cooljeanius) [reply data/core/units/elves/Champion.cfg:29]: for cross-referencing purposes, I opened #7895 for the `wmlunits` part

**Comment by cooljeanius** [comment]
> Movement changes are probably the changes most likely to need campaign changes. For example, SoF S07:
> 
> https://github.com/wesnoth/wesnoth/blob/1efa0c8c54eb9247b93bcc9dc1ba96cc477b183f/data/campaigns/Sceptre_of_Fire/scenarios/7_Outriding_the_Outriders.cfg#L324-L329
> 
> I don't know how many other examples there are, I've just remembered that scenario.

This was never addressed, and really should have been. Since it wasn't, it's now caused #8589

**Comment by spixi** [comment]
Regarding the Merman changes:

The difference between Netcaster to Entangler is 16 Gold for 54 XP
The difference between Spearman and Javelineer is 33 Gold for 54 XP

This is a hard mismatch. The Netcaster is more expensive than the Spearman, but it is the other way around for their Level 3 counterparts. Either, the XP requirement of the Spearmen has to be increased or the Javelineer needs a nerf and a cost reduction.

**Comment by spixi** [comment]
Suggestions:
* Increase the Entangler's melee attack from 8-3 to 9-3, because of RIPLIB
* Set the Entangler's cost back to 46
* Increase the Spearman's HP from 43 to 44 and its cost from 22 to 23
* Add parry 10 % to Spearman's and Javelineer's melee attack
* Increase XP requirement of the Spearman from 54 to 72 (with the slightly higher bulk it is easier to level up)

**Comment by spixi** [comment]
Javelineer at 55 gold is pretty expensive in comparision to Siren, which has the same MP and defenses, but a 15-3 magical ranged attack, compared to the 10-4 non-magical of the Javelineer AND only costs 42. 7 mote HP is not worth 13 gold.

**Comment by knyghtmare** [comment]
@spixi Please open a new discussional Issue ticket here instead of commenting on a merged pull request.
